# 🚀 Phase 2: Unsupervised Clustering of Engine Health Stages

This notebook focuses on applying unsupervised clustering techniques to the cleaned CMAPSS engine sensor datasets (FD001–FD004). The goal is to group engine cycles into **five degradation stages (Stage 0–Stage 4)**, which serve as interpretable, data-driven indicators of machinery health.

### 🎯 Why Clustering?
In real-world predictive maintenance, degradation stages are not always labeled. Instead of relying on arbitrary thresholds or synthetic labels, we apply clustering to let the **data speak for itself**. This allows us to:
- Detect **natural groupings** in sensor behavior across engine cycles.
- Define **progressive degradation stages** without bias or assumptions.
- Lay the foundation for future **classification or regression models** using these stages as targets.

### 🧠 Why Use Both KMeans and Agglomerative Clustering?
To ensure robust insights, we compare two fundamentally different unsupervised methods:
- **KMeans** assumes spherical, equally-sized clusters — useful for tight, centralized groupings.
- **Agglomerative Clustering** builds a hierarchy — excellent for uncovering gradual degradation paths.

Each offers complementary views of the underlying engine dynamics.

### 🔍 The Role of PCA and t-SNE
High-dimensional sensor data is projected into 2D using:
- **PCA (Principal Component Analysis)** to capture linear variance.
- **t-SNE (t-distributed Stochastic Neighbor Embedding)** to uncover non-linear patterns.

These visualizations help us qualitatively assess the separation and structure of clusters, offering intuitive insights into engine health transitions.

### 📊 Understanding Silhouette Scores
Silhouette score is a metric ranging from -1 to 1 that evaluates how well each data point fits into its assigned cluster. Higher scores indicate:
- Strong intra-cluster cohesion.
- Good inter-cluster separation.

We compute and compare silhouette scores for each method to objectively assess clustering quality across all four datasets.

### 🔄 What Comes Next?
The labeled stages generated here will act as **pseudo-labels** for:
- **Classification models** to predict the current health stage from live sensor readings.
- **Regression models** to estimate Remaining Useful Life (RUL) based on cluster-informed features.

This notebook is the bridge between unsupervised insight and supervised performance — and sets up the entire learning pipeline to follow.

In [ ]:
# Phase 2: Clustering Analysis with KMeans and AgglomerativeClustering (Enhanced + Plot Export + Score Summary)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans, AgglomerativeClustering
from sklearn.metrics import silhouette_score
import os

# Load cleaned data
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
os.makedirs(figures_path, exist_ok=True)

dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
datasets = {}

for ds_id in dataset_ids:
    df = pd.read_csv(os.path.join(base_path, f"clean_train_{ds_id}.csv"))
    datasets[ds_id] = df

# Map numeric cluster label to descriptive stage name
def stage_label(n):
    return f"Stage {n}"

# Global silhouette summary
silhouette_summary = {}

# Function to apply clustering and visualize with PCA and t-SNE
def apply_clustering(name, df):
    print(f"\n⏳ Clustering and visualizing {name}...")
    sensor_features = [col for col in df.columns if col.startswith("sensor_")]
    X = df[sensor_features].copy()

    # Apply clustering
    kmeans = KMeans(n_clusters=5, random_state=42)
    agglom = AgglomerativeClustering(n_clusters=5)
    df['kmeans_cluster'] = kmeans.fit_predict(X)
    df['agglo_cluster'] = agglom.fit_predict(X)
    df['kmeans_stage'] = df['kmeans_cluster'].apply(stage_label)
    df['agglo_stage'] = df['agglo_cluster'].apply(stage_label)

    # Silhouette scores
    k_score = silhouette_score(X, df['kmeans_cluster'])
    a_score = silhouette_score(X, df['agglo_cluster'])
    silhouette_summary[name] = {
        'KMeans': round(k_score, 4),
        'Agglomerative': round(a_score, 4)
    }

    # PCA visualization
    pca = PCA(n_components=2)
    pca_result = pca.fit_transform(X)
    df['pca_1'] = pca_result[:, 0]
    df['pca_2'] = pca_result[:, 1]

    plt.figure(figsize=(12, 5))
    plt.subplot(1, 2, 1)
    sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='kmeans_stage', palette='tab10')
    plt.title(f"{name}: KMeans Stages via PCA")
    plt.subplot(1, 2, 2)
    sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='agglo_stage', palette='tab10')
    plt.title(f"{name}: Agglomerative Stages via PCA")
    plt.tight_layout()
    plt.savefig(os.path.join(figures_path, f"{name}_PCA_Clusters.png"))
    plt.close()

    # t-SNE visualization (updated to use max_iter)
    tsne = TSNE(n_components=2, perplexity=30, max_iter=1000, random_state=42)
    tsne_result = tsne.fit_transform(X)
    df['tsne_1'] = tsne_result[:, 0]
    df['tsne_2'] = tsne_result[:, 1]

    plt.figure(figsize=(12, 5))
    plt.subplot(1, 2, 1)
    sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='kmeans_stage', palette='tab10')
    plt.title(f"{name}: KMeans Stages via t-SNE")
    plt.subplot(1, 2, 2)
    sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='agglo_stage', palette='tab10')
    plt.title(f"{name}: Agglomerative Stages via t-SNE")
    plt.tight_layout()
    plt.savefig(os.path.join(figures_path, f"{name}_TSNE_Clusters.png"))
    plt.close()

    # Distribution plots (added hue= explicitly to avoid warnings)
    plt.figure(figsize=(10, 4))
    plt.subplot(1, 2, 1)
    sns.countplot(x='kmeans_stage', hue='kmeans_stage', data=df, palette='tab10', order=sorted(df['kmeans_stage'].unique()))
    plt.title(f"{name} KMeans Stage Distribution")
    plt.subplot(1, 2, 2)
    sns.countplot(x='agglo_stage', hue='agglo_stage', data=df, palette='tab10', order=sorted(df['agglo_stage'].unique()))
    plt.title(f"{name} Agglomerative Stage Distribution")
    plt.tight_layout()
    plt.savefig(os.path.join(figures_path, f"{name}_Cluster_Distribution.png"))
    plt.close()

    # Save clustered data
    df.to_csv(os.path.join(base_path, f"clustered_train_{name}.csv"), index=False)
    print(f"✔ Clustered data saved: clustered_train_{name}.csv")

    # Clear and impactful explanation
    print(f"\n📘 **{name} Clustering Insight:**")
    print("Clustering the sensor data into degradation stages helped us visualize and interpret engine health patterns.")
    print("KMeans captured tight groupings while Agglomerative showed progressive transitions.")
    print("PCA and t-SNE plots make the separation between clusters clearly visible.")
    print("Silhouette scores validate the quality of the clustering for both methods.\n")

# Run clustering
for ds_id in dataset_ids:
    apply_clustering(ds_id, datasets[ds_id])
   

# Print silhouette summary for reporting
print("\n📊 **Silhouette Score Summary Table:**")
for ds, scores in silhouette_summary.items():
    print(f"{ds}: KMeans={scores['KMeans']}, Agglomerative={scores['Agglomerative']}")


## 📁 Save Silhouette Scores to JSON

This code saves the computed silhouette scores for all four datasets (FD001–FD004) into a `silhouette_scores.json` file.

**Why this matters:**  
- Silhouette scores are a standard way to evaluate how well clusters are formed.
- Saving them lets us reuse these results without recalculating them, especially when Phase 2 is too heavy to re-run.

**How it helps:**
- ✅ Validates our clustering quality in a reproducible way
- 📝 Makes it easy to include scores in the final report or paper
- 🔁 Prepares a reliable metric for comparing models later in Phase 5 (hybrid scoring)


In [6]:
import pandas as pd
import os
import json
from sklearn.metrics import silhouette_score

# Paths
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
report_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report'))
os.makedirs(report_path, exist_ok=True)

dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
silhouette_summary = {}

for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    if not os.path.exists(file_path):
        print(f"⚠️ File missing: {file_path}")
        continue

    df = pd.read_csv(file_path)
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]

    # Ensure required columns exist
    if 'kmeans_cluster' not in df.columns or 'agglo_cluster' not in df.columns:
        print(f"⚠️ Missing cluster labels in {ds_id}")
        continue

    # Compute silhouette scores
    X = df[sensor_cols]
    kmeans_score = silhouette_score(X, df['kmeans_cluster'])
    agglo_score = silhouette_score(X, df['agglo_cluster'])

    silhouette_summary[ds_id] = {
        'KMeans': round(kmeans_score, 4),
        'Agglomerative': round(agglo_score, 4)
    }

# Save to JSON
output_path = os.path.join(report_path, "silhouette_scores.json")
with open(output_path, "w") as f:
    json.dump(silhouette_summary, f, indent=4)

print(f"📁 Silhouette scores saved to: {output_path}")
print("📊 Summary:", silhouette_summary)


📁 Silhouette scores saved to: c:\Users\pandr\hybrid-predictive-maintenance\report\silhouette_scores.json
📊 Summary: {'FD001': {'KMeans': np.float64(0.188), 'Agglomerative': np.float64(0.176)}, 'FD002': {'KMeans': np.float64(0.884), 'Agglomerative': np.float64(0.884)}, 'FD003': {'KMeans': np.float64(0.2454), 'Agglomerative': np.float64(0.225)}, 'FD004': {'KMeans': np.float64(0.8862), 'Agglomerative': np.float64(0.8862)}}


## 📊 Silhouette Score Summary Table (KMeans vs Agglomerative)

This table summarizes the silhouette scores for each dataset and clustering method.  
Silhouette scores indicate how well-separated and cohesive the clusters are — higher is better.

| Dataset | KMeans Score | Agglomerative Score |
|---------|--------------|---------------------|
| FD001   | 0.1880       | 0.1760              |
| FD002   | 0.8840       | 0.8840              |
| FD003   | 0.2454       | 0.2250              |
| FD004   | 0.8862       | 0.8862              |

**Insights:**
- 🟢 FD002 and FD004 show excellent clustering quality (score > 0.88), suggesting clear degradation stages.
- 🟡 FD003 shows moderate cluster separation.
- 🔴 FD001 shows weak separation, possibly due to tighter degradation patterns or overlapping sensor signals.

These results help us decide which datasets are strong candidates for label-based supervised modeling in Phase 3 and Phase 5.


# 📄 Silhouette Markdown Summary Export
This cell converts the `silhouette_scores.json` file into a human-readable markdown file (`silhouette_summary.md`) inside the `/report/` folder. 
This helps us include the scores directly in documentation, papers, or GitHub views.


In [8]:
import os
import json

# Define report path
report_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report'))
json_path = os.path.join(report_path, "silhouette_scores.json")
summary_md = os.path.join(report_path, "silhouette_summary.md")

# Load JSON
with open(json_path, "r") as f:
    silhouette_summary = json.load(f)

# Write Markdown summary
with open(summary_md, "w", encoding="utf-8") as md:
    md.write("# 📊 Silhouette Score Summary\n\n")
    for ds, scores in silhouette_summary.items():
        md.write(f"## {ds}\n")
        md.write(f"- **KMeans**: {scores['KMeans']}\n")
        md.write(f"- **Agglomerative**: {scores['Agglomerative']}\n\n")


print(f"✅ Markdown summary saved to: {summary_md}")


✅ Markdown summary saved to: c:\Users\pandr\hybrid-predictive-maintenance\report\silhouette_summary.md


## 🔍 Top 5 Sensors by Variance (Per Dataset)

This quick EDA reveals which sensor features show the most variation.

**Why this matters:**  
- High variance = higher impact in clustering  
- Helps justify sensor selection for classifiers in Phase 3  
- Provides interpretable insights for final report


In [13]:
import json

top_variance_summary = {}

for ds_id in dataset_ids:
    df = pd.read_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"))
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    top_sensors = df[sensor_cols].var().sort_values(ascending=False).head(5).round(4)
    
    # Store as dict
    top_variance_summary[ds_id] = top_sensors.to_dict()

    # Print in notebook
    print(f"\n📊 {ds_id} - Top 5 High-Variance Sensors:")
    print(top_sensors)

# Save to file
os.makedirs("../report", exist_ok=True)
with open("../report/top_variance_sensors.json", "w") as f:
    json.dump(top_variance_summary, f, indent=4)

print("\n📁 Saved top variance sensor summary to: report/top_variance_sensors.json")



📊 FD001 - Top 5 High-Variance Sensors:
sensor_11    0.0253
sensor_12    0.0247
sensor_4     0.0231
sensor_2     0.0227
sensor_21    0.0223
dtype: float64

📊 FD002 - Top 5 High-Variance Sensors:
sensor_16    0.2220
sensor_1     0.1283
sensor_19    0.1267
sensor_13    0.1246
sensor_12    0.1219
dtype: float64

📊 FD003 - Top 5 High-Variance Sensors:
sensor_11    0.0294
sensor_12    0.0275
sensor_7     0.0271
sensor_17    0.0256
sensor_4     0.0232
dtype: float64

📊 FD004 - Top 5 High-Variance Sensors:
sensor_16    0.2195
sensor_1     0.1288
sensor_19    0.1269
sensor_13    0.1248
sensor_2     0.1175
dtype: float64

📁 Saved top variance sensor summary to: report/top_variance_sensors.json


## ⏱️ Average Engine Cycle per Cluster (Degradation Profile)

This plot shows the **average time (cycle count)** at which each cluster/stage appears in the dataset.

**Why this matters:**  
- A good degradation model should show Stage 0 early in life, and Stage 4 closer to failure.
- This confirms that clusters follow a **progressive degradation pattern**.

**How it helps:**
- ✅ Validates that our unsupervised labels align with real-world time progression
- 🧩 Supports hybrid modeling by linking clusters to RUL logic
- 📉 Adds an insightful diagnostic to the final report


In [6]:
import matplotlib.pyplot as plt
for ds_id in dataset_ids:
    df = pd.read_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"))
    df['time'] = df['time'] if 'time' in df.columns else df.index + 1

    plt.figure(figsize=(6, 4))
    mean_time = df.groupby('kmeans_stage')['time'].mean().sort_index()
    mean_time.plot(kind='bar', color='skyblue')
    plt.title(f"{ds_id} - Avg Cycle by KMeans Stage")
    plt.ylabel("Average Time (Cycle)")
    plt.tight_layout()
    plt.savefig(f"../figures/{ds_id}_Stage_TimeProfile.png")
    plt.close()

## 📋 Cluster Summary Table (Sensor Means by Stage)

This prints the average sensor values and time per stage.

**Why this matters:**  
- Adds interpretability to clusters  
- Shows what “Stage 0” or “Stage 4” actually means  
- Useful to explain degradation patterns in Phase 5


In [14]:
summary_folder = os.path.abspath(os.path.join(os.getcwd(), '..', 'report', 'cluster_summaries'))
os.makedirs(summary_folder, exist_ok=True)

for ds_id in dataset_ids:
    df = pd.read_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"))
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    
    # Include time column if available
    cols_to_use = sensor_cols + ['time'] if 'time' in df.columns else sensor_cols
    summary = df.groupby("kmeans_stage")[cols_to_use].mean().round(3)
    
    # Save as CSV
    summary.to_csv(os.path.join(summary_folder, f"{ds_id}_ClusterSummary.csv"))
    print(f"📁 Saved: {ds_id}_ClusterSummary.csv")


📁 Saved: FD001_ClusterSummary.csv
📁 Saved: FD002_ClusterSummary.csv
📁 Saved: FD003_ClusterSummary.csv
📁 Saved: FD004_ClusterSummary.csv


## 📈 Elbow Curves for All Datasets (KMeans)

This code plots the WCSS (within-cluster sum of squares) for K values from 2 to 10 for each dataset.

**Why this matters:**  
- Helps validate the choice of 5 clusters  
- Shows how compactness improves with more clusters  
- Supports model justification in final report


In [11]:
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

for ds_id in dataset_ids:
    df = pd.read_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"))
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    X = df[sensor_cols]

    wcss = []
    for k in range(2, 11):
        model = KMeans(n_clusters=k, random_state=42)
        model.fit(X)
        wcss.append(model.inertia_)

    plt.figure(figsize=(6, 4))
    plt.plot(range(2, 11), wcss, marker='o', color='green')
    plt.title(f"Elbow Curve for KMeans - {ds_id}")
    plt.xlabel("Number of Clusters (k)")
    plt.ylabel("WCSS")
    plt.grid(True)
    plt.tight_layout()
    plt.savefig(f"../figures/{ds_id}_ElbowCurve.png")
    plt.close()


## 🌳 Dendrograms for Agglomerative Clustering (All Datasets)

This plots a dendrogram using Ward linkage on a random 500-sample subset for each dataset.

**Why this matters:**  
- Reveals hierarchical structure in clustering  
- Useful to understand how engines degrade in stages  
- Adds depth to Agglomerative Clustering results in Phase 2


In [12]:
from scipy.cluster.hierarchy import linkage, dendrogram

for ds_id in dataset_ids:
    df = pd.read_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"))
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    X_sample = df[sensor_cols].sample(n=500, random_state=42)

    linkage_matrix = linkage(X_sample, method='ward')

    plt.figure(figsize=(10, 4))
    dendrogram(linkage_matrix, truncate_mode='lastp', p=20)
    plt.title(f"Dendrogram - {ds_id} (Sample of 500)")
    plt.tight_layout()
    plt.savefig(f"../figures/{ds_id}_Dendrogram.png")
    plt.close()


## 📊 Summary Plot: Top 3 High-Variance Sensors per Cluster Stage (KMeans)

This section visualizes how the top 3 high-variance sensors behave across the five KMeans cluster stages (Stage 0–4) for each FD dataset.

### 🔍 Why this is important:
- Helps identify whether the **most variable sensors** align well with cluster transitions.
- Provides **stage-wise sensor pattern insight**, which is crucial for **feature selection in Phase 3 (classification)**.
- Enhances interpretability — shows which sensors are stable, rising, or degrading in each stage.
- Strengthens clustering validation beyond silhouette scores and PCA/t-SNE visuals.

### 📈 What this plot shows:
For each dataset (FD001–FD004), the average value of the top 3 highest variance sensors is plotted **per KMeans stage**, based on the clustered training data.

This helps answer: *Do the most sensitive sensors meaningfully separate stages?*

The plots are automatically saved in the `/figures/` folder with filenames like:



In [9]:
# 📊 Top Sensor Summary Plots per Stage (FD001–FD004)
import os
import json
import pandas as pd
import matplotlib.pyplot as plt

# Paths
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
fig_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
json_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report', 'top_variance_sensors.json'))

# Load top sensors JSON
with open(json_path, "r") as f:
    top_sensors = json.load(f)

# Create and save plots
for ds_id, sensors in top_sensors.items():
    file = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    if not os.path.exists(file):
        print(f"❌ File not found: {file}")
        continue

    df = pd.read_csv(file)
    if "kmeans_stage" not in df.columns:
        print(f"❌ Missing clustering column in: {ds_id}")
        continue

    top3 = list(sensors.keys())[:3]  # Take top 3 sensors
    mean_vals = df.groupby("kmeans_stage")[top3].mean()

    # Plot
    mean_vals.plot(kind="bar", figsize=(8, 5))
    plt.title(f"{ds_id} – Top 3 Sensor Averages by Stage (KMeans)")
    plt.ylabel("Average Normalized Sensor Value")
    plt.xlabel("KMeans Stage")
    plt.xticks(rotation=0)
    plt.tight_layout()

    save_path = os.path.join(fig_path, f"{ds_id}_TopSensors_PerStage.png")
    plt.savefig(save_path)
    plt.close()
    print(f"✅ Saved: {save_path}")


✅ Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD001_TopSensors_PerStage.png
✅ Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD002_TopSensors_PerStage.png
✅ Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD003_TopSensors_PerStage.png
✅ Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD004_TopSensors_PerStage.png
